tools/PI/DevHome.PI/Controls/AddToolControl.xaml.cs

```diff
@@ -101,55 +161,290 @@ private void ClearValues()
+            LoadingProgressTextRing.Value = percent;
+        });
+
+        await ProcessItemsAsync(progress);
+
+        foreach (var app in allApps)
+        {
+            SortedApps.Add(app);
+        }
+
+        IsLoading = false;
+    }
+
+    private int GetShortcuts()
+    {
+        var count = 0;
+
+        // Search for .lnk files in the per-user and all-users Start Menu Programs directories.
+        // %APPDATA%\Microsoft\Windows\Start Menu\Programs
+        var startMenuProgramsPath =
+            Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.StartMenu), "Programs");
+
+        // %ProgramData%\Microsoft\Windows\Start Menu\Programs
+        var commonStartMenuProgramsPath =
+            Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.CommonStartMenu), "Programs");
+
+        shortcuts = [];
+        if (Directory.Exists(startMenuProgramsPath))
+        {
+            shortcuts.AddRange(Directory.GetFiles(startMenuProgramsPath, "*.lnk", SearchOption.AllDirectories));
+        }
+
+        if (Directory.Exists(commonStartMenuProgramsPath))
+        {
+            shortcuts.AddRange(Directory.GetFiles(commonStartMenuProgramsPath, "*.lnk", SearchOption.AllDirectories));
+        }
+
+        if (shortcuts is not null)
+        {
+            count = shortcuts.Count;
+        }
+
+        return count;
     }
 
-    private void UpdateSampleCommandline()
+    private int GetPackages()
     {
-        if (SampleCommandTextBox is null)
+        var count = 0;
+        var packageManager = new PackageManager();
+        packages = packageManager.FindPackagesForUserWithPackageTypes(string.Empty, PackageTypes.Main).ToList();
+        if (packages is not null)
+        {
+            count = packages.Count;
+        }
+
+        return count;
+    }
+
+    private async Task ProcessItemsAsync(IProgress<int> progress)
+    {
+        // Recursively search for .lnk files.
+        var currentCount = 0;
+        for (var i = 0; i < shortcuts?.Count; i++)
+        {
+            await Task.Run(() => ProcessShortcut(shortcuts[i]));
+
+            // Report progress.
+            currentCount++;
+            var percentComplete = (i + 1) * ShortcutProcessingEndIndex / itemCount;
+            progress.Report(percentComplete);
+        }
+
+        for (var j = 0; j < packages?.Count; j++)
+        {
+            await Task.Run(() => ProcessPackage(packages[j]));
+
+            // Report progress.
+            currentCount++;
+            var percentComplete = ShortcutProcessingEndIndex + ((j + 1) * ShortcutProcessingEndIndex / itemCount);
+            progress.Report(percentComplete);
+        }
+
+        allApps = allApps.OrderBy(app => app.Name).ToList();
+
+        // We get the icon data on the UI thread, because BitmapImages must be created on the UI thread.
+        for (var k = 0; k < allApps.Count; k++)
+        {
+            var app = allApps[k];
+            if (app.IsMsix)
```

Comment:
Prefer early continues over a deep if nest.

Reply:
Skipping any items with continue would be wrong: we have 2 different behaviors based on the if condition.